### PR TITLE
deployment controller: cancel deployers on new cancelled deployments

### PR DESCRIPTION
**Pull request by kargakis**
Deployer pods may stay around due to an unavailable node and get scheduled
after a cancelled deployment has been transitioned from New to Failed.

Found out in https://github.com/openshift/origin/issues/8403

@pweil- @smarterclayton @ironcladlou

### CODE REVIEW

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/util/util.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func LabelForDeployment(deployment *api.ReplicationController) string {
 
 // LabelForDeploymentConfig builds a string identifier for a DeploymentConfig.
 func LabelForDeploymentConfig(config *deployapi.DeploymentConfig) string {
-	return fmt.Sprintf("%s/%s:%d", config.Namespace, config.Name, config.Status.LatestVersion)
+	return fmt.Sprintf("%s/%s", config.Namespace, config.Name)
```
Comment (soltysh) [inline pkg/deploy/util/util.go]: Was this change intentional?

File: pkg/deploy/controller/deployment/controller.go [hunk 2]
```diff
@@ -59,6 +59,9 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 		// transition to failed immediately.
 		if deployutil.IsDeploymentCancelled(deployment) {
 			nextStatus = deployapi.DeploymentStatusFailed
+			if err := c.cancelDeployerPods(deployment); err != nil {
+				return err
```
Comment (smarterclayton) [inline pkg/deploy/controller/deployment/controller.go]: Maybe a larger question - when we cancel, why don't we update the RC and then delete the deployer pods?  Is that going to race with the deployment process pod controller?  Or are you just trying to ensure it happens before and after?

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

File: pkg/deploy/controller/deployment/controller.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/deploy/controller/deployment/controller.go]: Interesting q... I do think that's a larger investigation. Pretty sure this patch will do what we want for now.

**Comment by ironcladlou** [comment]
LGTM once https://github.com/openshift/origin/pull/8418#discussion_r59013279 is answered

**Comment by smarterclayton** [comment]
Can we spawn a follow up to track the bigger implications of this (in the
context of the even larger changes to deployments going on)?

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@soltysh has the failure for pullthrough

On Apr 8, 2016, at 3:50 PM, OpenShift Bot <notifications@github.com> wrote:

continuous-integration/openshift-jenkins/test FAILURE (
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2856/)

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8418#issuecomment-207575875>

File: pkg/deploy/util/util.go (hunk 1 above)
Comment (kargakis) [inline pkg/deploy/util/util.go]: Yes, this is confusing since we already name deployments after their version.

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/controller/deployment/controller.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/controller/deployment/controller.go]: Spawned https://github.com/openshift/origin/issues/8439

**Comment by smarterclayton** [comment]
Approved [merge]

**Comment by kargakis** [comment]
https://github.com/openshift/origin/issues/7429 flake

re[merge]

File: pkg/deploy/util/util.go (hunk 1 above)
Comment (soltysh) [inline pkg/deploy/util/util.go]: Isn't this going to cause backwards compatibility issues?

Comment (kargakis) [inline pkg/deploy/util/util.go]: Why? AFAIK we use it only for logging

Comment (soltysh) [inline pkg/deploy/util/util.go]: Yup, my grep-fuu failed me.
